### PR TITLE
Load config from `~/rubocop/` and `$XDG_CONFIG_HOME`

### PR DESCRIPTION
**Pull request by tejasbubane**
if `$HOME` not set

Spec: https://specifications.freedesktop.org/basedir-spec/basedir-spec-0.8.html

Closes: #6662

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

### CODE REVIEW

**Comment by Mange** [comment]
In which context would `$HOME` not be set but a `$XDG_CONFIG_HOME` is set?

I went here to see if there was an issue for XDG support already. My take on it would be this:

- Check for `$XDG_CONFIG_HOME/rubocop/config.yml` and load it if it exists, else go to next step
- Check for `$HOME/.rubocop.yml` and load it if it exits.

Then existing users of the dotfile would not see any breakages, but XDG-focused users could move their existing file to the correct location and have it load automatically.

**Comment by tejasbubane** [comment]
@Mange Gotcha, I did it the other way round. You are right, I will change the implementation. Thanks!

**Comment by tejasbubane** [comment]
@Mange Fixed. I made some changes, this is the order being checked now:

```
RuboCop::ConfigLoader
  .configuration_file_for
    when a config file exists in the parent directory
      returns the path to that configuration file
    when multiple config files exist in ancestor directories
      prefers closer config file
    when no config file exists in ancestor directories
      ENV has `XDG_HOME` defined
        config file exists in `XDG_HOME` directory
          returns path to the file in `XDG_HOME` directory
        no config file in `XDG_HOME` but in rubocop directory
          returns path to the file in XDG_HOME/rubocop directory
      when ENV does not have `XDG_HOME` defined
        and a config file exists in home directory
          returns the path to the file in home directory
        and a config file exists in rubocop directory
          returns the path to the file in home/rubocop directory
      and no config file exists in `XDG_HOME` or `HOME` directories
        falls back to the provided default file
```

**Comment by Mange** [comment]
I cannot find any references to a `XDG_HOME` directory in the specification: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

Is it from some other specification, or is it a mistake? Are you mixing up XDG_CONFIG_HOME with HOME?

Normal config path according to the specification would be `$XDG_CONFIG_HOME/rubocop/config.yml`, aka `$HOME/.config/rubocop/config.yml` (if XDG_CONFIG_HOME isn't defined).

**Comment by tejasbubane** [comment]
@Mange That was indeed a mistake. I changed it to `XDG_CONFIG_HOME`.

**Comment by Mange** [comment]
I think this looks good now overall. I hope the maintainers agree.

File: lib/rubocop/file_finder.rb [hunk 1]
```diff
@@ -31,17 +31,26 @@ def find_files_upwards(filename, start_dir, use_home: false)
     private
 
     def traverse_files_upwards(filename, start_dir, use_home)
+      traverse_file(filename, start_dir).each { |f| yield(f.to_s) }
+      return unless use_home && (ENV.key?('HOME') ||
```
Comment (bbatsov) [inline lib/rubocop/file_finder.rb]: Seems to me that we're mixing different concerns in this method. Ideally the file traversal should be separate from checks for which directory contains the config file.

File: lib/rubocop/config_loader.rb [hunk 2]
```diff
@@ -15,6 +15,7 @@ class ConfigNotFoundError < Error
   # directories are inspected.
   class ConfigLoader
     DOTFILE = '.rubocop.yml'.freeze
+    CONFIGFILE = 'rubocop/config.yml'.freeze
```
Comment (bbatsov) [inline lib/rubocop/config_loader.rb]: Probably this should be `.rubocop`, not `rubocop`. Usually config dirs are "hidden". I'd also name the constant `CONFIG_FILE`.

**Comment by bbatsov** [comment]
I left a few small comments. Generally I'm fine with the proposed solution.

File: lib/rubocop/config_loader.rb (hunk 2 above)
Comment (Mange) [reply lib/rubocop/config_loader.rb]: Files in XDG_CONFIG_HOME should not be hidden according to specs. The most common convention is `appname/config`.

**Comment by bbatsov** [comment]
I'm closing this PR due to no recent activity. Feel free to re-open it if you ever come back to it.

**Comment by Mange** [comment]
I re-opened this in #6895.

**Pull request closed without merging**